addons/stock/models/product.py

```diff
@@ -906,12 +923,6 @@ def write(self, vals):
                 if quant:
                     raise UserError(_("This product's company cannot be changed as long as there are quantities of it belonging to another company."))
 
-        if 'uom_id' in vals:
-            new_uom = self.env['uom.uom'].browse(vals['uom_id'])
-            updated = self.filtered(lambda template: template.uom_id != new_uom)
-            done_moves = self.env['stock.move'].sudo().search([('product_id', 'in', updated.with_context(active_test=False).mapped('product_variant_ids').ids)], limit=1)
-            if done_moves:
-                raise UserError(_("You cannot change the unit of measure as there are already stock moves for this product. If you want to change the unit of measure, you should rather archive this product and create a new one."))
```

Comment:
I see the justification for this is to make it easier to fix mis-configurations, but the "they will have to do a manual adjustment to fix it" isn't that simple :cry:. We will need to check that the quant reservation logic is still correct (i.e. On Hand/Reserved Qtys). Not sure if we will have to worry about inconsistent db states too if someone does something like validate a picking right when the UoM is changed

Reply:
Just kidding, we can do something similar to the `uom.uom` check:

https://github.com/odoo/odoo/blob/4bda75cc8cba48ce0ce0351cbf27aad1f00f2b10/addons/stock/models/product.py#L1125-L1134

and ensure that you cannot change the UoM if there are non-done, i.e. potentially reserved moves (i.e. `state not in [draft, cancel, done]`). This will ensure that we don't have incorrect reserved/availability qtys. We should also make the UoM tracked so that users can match up between done moves and when the uom was changed so they can track when the numbers don't necessarily make sense